features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/vitals/FPSVitalListener.kt

```diff
@@ -0,0 +1,58 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+package com.datadog.android.rum.internal.vitals
+
+import android.os.Build
+import androidx.metrics.performance.FrameData
+import com.datadog.android.core.internal.system.BuildSdkVersionProvider
+import com.datadog.android.rum.internal.domain.FrameMetricsData
+import java.util.concurrent.TimeUnit
+
+internal class FPSVitalListener(
+    private val vitalObserver: VitalObserver,
+    private val buildSdkVersionProvider: BuildSdkVersionProvider = BuildSdkVersionProvider.DEFAULT,
+    private var screenRefreshRate: Double = 60.0
+) : FrameStateListener {
+    private var frameDeadline = SIXTEEN_MS_NS
+    private var displayRefreshRate: Double = SIXTY_FPS
+
+    override fun onFrame(volatileFrameData: FrameData) {
+        val durationNs = volatileFrameData.frameDurationUiNanos
+        if (durationNs > 0.0) {
+            var frameRate = (ONE_SECOND_NS / durationNs)
+
+            if (buildSdkVersionProvider.version >= Build.VERSION_CODES.S) {
+                screenRefreshRate = ONE_SECOND_NS / frameDeadline
+            } else if (buildSdkVersionProvider.version == Build.VERSION_CODES.R) {
+                screenRefreshRate = displayRefreshRate
+            }
+
+            // If normalized frame rate is still at over 60fps it means the frame rendered
+            // quickly enough for the devices refresh rate.
+            frameRate = (frameRate * (SIXTY_FPS / screenRefreshRate)).coerceAtMost(MAX_FPS)
+
+            if (frameRate > MIN_FPS) {
+                vitalObserver.onNewSample(frameRate)
+            }
+        }
+    }
+
+    override fun onFrameMetricsData(data: FrameMetricsData) {
+        displayRefreshRate = data.displayRefreshRate
+        if (buildSdkVersionProvider.version >= Build.VERSION_CODES.S) {
+            frameDeadline = data.deadline
+        }
+    }
+
+    companion object {
+        private const val SIXTEEN_MS_NS: Long = 16666666
```

Comment:
This value looks a bit strange to me. Why not something like 16_000_000L?

Reply:
this is actually the result of `1/60`, which is 16.(6).